pkg/node/types/node.go

```diff
@@ -303,6 +307,19 @@ func (n *Node) GetCiliumInternalIP(ipv6 bool) net.IP {
 	return nil
 }
 
+func (n *Node) GetIPByType(addrType addressing.AddressType, ipv6 bool) net.IP {
+	for _, addr := range n.IPAddresses {
+		if (ipv6 && addr.IP.To4() != nil) ||
+			(!ipv6 && addr.IP.To4() == nil) {
+			continue
+		}
+		if addr.Type == addrType {
+			return addr.IP
+		}
+	}
```

Comment:
Move the address type check before the IP check, as `net.IP.To4()` might allocate:

```suggestion
	for _, addr := range n.IPAddresses {
                if addr.Type != addrType {
                        continue
                }
                if is4 := addr.IP.To4() != nil; (!ipv6 && is4) || (ipv6 && !is4) {
                        return addr.IP
                }
	}
```

FWIW, I've verified the code above using the following unit test which passes both with and without the change:

<details>
  <summary>Diff for <code>(*Node).GetIPByType</code> unit test</summary>

```diff
diff --git a/pkg/node/types/node_test.go b/pkg/node/types/node_test.go
index 6a16d5c3808c..c11e8bfd51a2 100644
--- a/pkg/node/types/node_test.go
+++ b/pkg/node/types/node_test.go
@@ -78,6 +78,60 @@ func (s *NodeSuite) TestGetNodeIP(c *C) {
 
 }
 
+func (s *NodeSuite) TestGetIPByType(c *C) {
+       n := Node{
+               Name: "node-1",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("192.0.2.3"), Type: addressing.NodeExternalIP},
+               },
+       }
+
+       ip := n.GetIPByType(addressing.NodeInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeInternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip.Equal(net.ParseIP("192.0.2.3")), Equals, true)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip, IsNil)
+
+       n = Node{
+               Name: "node-2",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("f00b::1"), Type: addressing.NodeCiliumInternalIP},
+               },
+       }
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeCiliumInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeCiliumInternalIP, true)
+       c.Assert(ip.Equal(net.ParseIP("f00b::1")), Equals, true)
+
+       n = Node{
+               Name: "node-3",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("192.42.0.3"), Type: addressing.NodeExternalIP},
+                       {IP: net.ParseIP("f00d::1"), Type: addressing.NodeExternalIP},
+               },
+       }
+
+       ip = n.GetIPByType(addressing.NodeInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeInternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip.Equal(net.ParseIP("192.42.0.3")), Equals, true)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip.Equal(net.ParseIP("f00d::1")), Equals, true)
+}
+
```

</details>

Feel free to pick this up into the respective commit.

Reply:
Very nice suggestion, thanks a lot! Will apply in the next push!